cmd/kubeadm/app/cmd/phases/workflow/runner.go

```diff
@@ -0,0 +1,382 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package workflow
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+)
+
+// phaseSeparator defines the separator to be used when concatenating nested
+// phase names
+const phaseSeparator = "/"
+
+// RunnerOptions defines the options supported during the execution of a
+// kubeadm composable workflows
+type RunnerOptions struct {
+	// PhaseFilter defines the list of phases to be executed (if empty, all).
+	PhaseFilter []string
+
+	// ExcludePhases defines the list of phases to be excluded by execution (if empty, none).
+	ExcludePhases []string
+}
+
+// RunnerContext defines the type of runner context, that is any type.
+type RunnerContext = interface{}
+
+// Runner provides a simple, lightweight runner for composable kubeadm workflows.
+type Runner struct {
+	// Options that regulate the runner behavior.
+	Options RunnerOptions
+
+	// Phases composing the workflow to be managed by the runner.
+	Phases []Phase
+
+	// contextBuilder defines a function that creates the runtime context shared
+	// among all the phases included in the workflow
+	contextBuilder func() (RunnerContext, error)
+
+	ctx RunnerContext
+
+	// phaseRunners is part of the internal state of the runner and provides
+	// a list of wrappers to phases composing the workflow, providing contextual
+	// information to the phases to be executed.
+	phaseRunners []*phaseRunner
+}
+
+// phaseRunner provides a wrapper to a Phase with the addition of a set
+// of contextual information derived by the workflow managed by the Runner.
+type phaseRunner struct {
+	// Phase provide access to the phase implementation
+	Phase
+
+	// provide access to the parent phase in the workflow managed by the Runner.
+	parent *phaseRunner
+
+	// level define the level of nesting of this phase into the workflow managed by
+	// the Runner.
+	level int
```

Comment:
Phase runner supports this feature because currently there are nested phases and I think we should reproduce it.
I agree that this feature should not be abused when implementing phases.